scipy/signal/tests/test_ltisys.py

```diff
@@ -8,11 +8,123 @@
 from scipy.signal.ltisys import (ss2tf, tf2ss, lsim2, impulse2, step2, lti,
                                  bode, freqresp, lsim, impulse, step,
                                  abcd_normalize, place_poles,
-                                 TransferFunction, StateSpace, ZerosPolesGain)
+                                 TransferFunction, StateSpace, ZerosPolesGain,
+                                 controllability_matrix,
+                                 observability_matrix, controller_hessenberg)
+
 from scipy.signal.filter_design import BadCoefficients
 import scipy.linalg as linalg
 
 
+class TestCtrbObsv:
+    # The output from controller_hessenberg has been tested with random
+    # controllable and uncontrollable system against Octave which uses tb01ud
+    # from Slicot and all results were similar using the default tolerances of
+    # `allclose`.
+
+    def test_hessenberg(self):
+        # use np.allclose tolerances, atol = 0 from assert_allclose is
+        # unrealistic here especially for the comparison between
+        # np.linalg.inv(h_f.P) and h_f.P.T
+
+        # Test from the book the algorithm comes from (see [1] in docstring)
+        # it gives us references values to test against what we compute (p 177)
+        A = np.array([[1, 1, 1], [1, 1, 1], [0, 0, 1]])
+        B = np.array([[1, 1], [1, 1], [1, 1]])
+
+        ref_Abar = np.array([[2.3333, -0.4714, 0], [0.9428, 0.6667, 0],
+                             [0, 0, 0]])
+        ref_Bbar = np.array([[-1.7321, -1.7321], [0, 0], [0, 0]])
+
+        h_f = controller_hessenberg(A, B)
+
+        assert_allclose(h_f.P.dot(A.dot(h_f.P.T)), h_f.Abar,
+                        rtol=1e-5, atol=1e-8)
+        assert_allclose(h_f.P.dot(B), h_f.Bbar, rtol=1e-5, atol=1e-8)
+        assert_allclose(np.linalg.inv(h_f.P), h_f.P.T, rtol=1e-5, atol=1e-8)
+        assert_equal(2, h_f.control_index)
+        # the reference book only gives 5 digits, we have to relax allclose
+        assert_allclose(h_f.Abar, ref_Abar, rtol=1e-5, atol=1e-4)
+        assert_allclose(h_f.Bbar, ref_Bbar, rtol=1e-5, atol=1e-4)
+
+        # This one is from the reference book also but interestingly the final
+        # result shown page 181 is neither what we get nor what octave outputs.
+        # Let's check against Octave as more throughout tests showed that
+        # controller_hessenberg always has the same output.
+
+        A = np.array([0.7665, 0.1665, 0.9047, 0.4540, 0.5007, 0.4777, 0.4865,
+                      0.5045, 0.2661, 0.3841, 0.2378, 0.8977, 0.5163, 0.0907,
+                      0.2771, 0.2749, 0.9092, 0.3190, 0.9478, 0.9138, 0.3593,
+                      0.0606, 0.9866, 0.0737, 0.5297]).reshape((5, 5))
```

Comment:
I like to input big arrays like that, this way I don't have to wonder where to put the [] stuff, all I need is a reshape. Admittedly  it is less readable but I tend to think that if someone really needs to check this array all he has to do is to download the original publication and he will see it in a much more readable way.
However you're right about the array() before trying ndim, I'll fix that.